psbt/finalizer.go

```diff
@@ -0,0 +1,452 @@
+// has all partial signatures and redeem scripts/witness scripts
+// already prepared for the specified input, and so removes all temporary
+// data and replaces them with completed scriptSig and witness
+// fields, which are stored in key-types 07 and 08. The witness/
+// non-witness utxo fields in the inputs (key-types 00 and 01) are
+// left intact as they may be needed for validation (?).
+// If there is any invalid or incomplete data, an error is
+// returned.
+func Finalize(p *Psbt, idx int) error {
+	var err error
+	pInput := p.Inputs[idx]
+	if pInput.WitnessUtxo != nil {
+		err = FinalizeWitness(p, idx)
+		if err != nil {
+			return err
+		}
+	} else if pInput.NonWitnessUtxo != nil {
+		err = FinalizeNonWitness(p, idx)
+		if err != nil {
+			return err
+		}
+	} else {
+		return ErrInvalidPsbtFormat
+	}
+
+	if err = p.SanityCheck(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// checkFinalScriptSigWitness checks whether a given input in the
+// Psbt struct already has the fields 07 (FinalInScriptSig) or 08
+// (FinalInWitness). If so, it returns true. It does not modify the
+// Psbt.
+func checkFinalScriptSigWitness(p *Psbt, idx int) bool {
+	pInput := p.Inputs[idx]
+	if pInput.FinalScriptSig != nil {
+		return true
+	}
+	if pInput.FinalScriptWitness != nil {
+		return true
+	}
+	return false
+}
+
+// FinalizeNonWitness attempts to create PsbtInFinalScriptSig field
+// for input at index inIndex, and removes all other fields except
+// for the utxo field, for an input of type non-witness, or returns
+// an error.
+func FinalizeNonWitness(p *Psbt, inIndex int) error {
+	if checkFinalScriptSigWitness(p, inIndex) {
+		return ErrInputAlreadyFinalized
+	}
+	// Construct a scriptSig given the pubkey, signature (keytype 02),
+	// of which there might be multiple, and the redeem script
+	// field (keytype 04) if present (note, it is not present
+	// for p2pkh type inputs).
+	var scriptSig []byte
+	var err error
+	pInput := p.Inputs[inIndex]
+	containsRedeemScript := pInput.RedeemScript != nil
+	var pubKeys [][]byte
+	var sigs [][]byte
+	for _, ps := range pInput.PartialSigs {
+		pubKeys = append(pubKeys, ps.PubKey)
+		sigOK := checkSigHashFlags(ps.Signature, &pInput)
+		if !sigOK {
+			return ErrInvalidSigHashFlags
+		}
+		sigs = append(sigs, ps.Signature)
+	}
+
+	if len(sigs) < 1 || len(pubKeys) < 1 {
+		// We have failed to identify at least 1 (sig, pub) pair
+		// in the PSBT, which indicates it was not ready to be finalized.
+		return ErrNotFinalizable
+	}
+
+	if !containsRedeemScript {
+		// p2pkh
+		builder := txscript.NewScriptBuilder()
+		builder.AddData(sigs[0]).AddData(pubKeys[0])
+		scriptSig, err = builder.Script()
+		if err != nil {
+			return err
+		}
+	} else {
+		// This is assumed p2sh multisig
+		// Given redeemScript and pubKeys we can decide in what order
+		// signatures must be appended.
+		orderedSigs, err := extractKeyOrderFromScript(pInput.RedeemScript,
+			pubKeys, sigs)
+		if err != nil {
```

Comment:
Yes, I like/agree with your points here. Obviously only supporting multisig is a bit crappy, I think perhaps refactor a bit so at least there is a "not implemented" function to support generic Script rather than msig only, which can be added later, and also your point about checking by popping off the stack makes a lot of sense. I will try to do that part at least.

Reply:
Partially done in 1d79a3e2102b4c8b4929d0f446de18b65453f03f - will next look at the M count (ideally parse it properly but not clear if what's needed is exported, we'll see)

Reply:
Done in 021250ec3443533a7e9139f7ad1580488489bcf3 - see comment below.